src/StreamJsonRpc/Reflection/TargetMethod.cs

```diff
@@ -100,80 +102,47 @@ private void AddErrorMessage(string message)
-
-                var args = new object[includeCancellationToken ? 2 : 1];
-                args[0] = request.Parameters;
-                if (includeCancellationToken)
-                {
-                    args[1] = CancellationToken.None;
-                }
-
-                return args;
+                arguments[arguments.Length - 1] = CancellationToken.None;
             }
 
-            // The number of parameters must fall within required and total parameters.
-            int paramCount = request.ParameterCount;
-            if (paramCount < method.RequiredParamCount || paramCount > method.TotalParamCountExcludingCancellationToken)
+            switch (request.TryGetTypedArguments(methodParametersExcludingCancellationToken, argumentsExcludingCancellationToken))
             {
-                string methodParameterCount;
-                if (method.RequiredParamCount == method.TotalParamCountExcludingCancellationToken)
-                {
-                    methodParameterCount = method.RequiredParamCount.ToString(CultureInfo.CurrentCulture);
-                }
-                else
-                {
+                case JsonRpcRequest.ArgumentMatchResult.Success:
+                    return true;
+                case JsonRpcRequest.ArgumentMatchResult.ParameterArgumentCountMismatch:
+                    string methodParameterCount;
                     methodParameterCount = string.Format(CultureInfo.CurrentCulture, "{0} - {1}", method.RequiredParamCount, method.TotalParamCountExcludingCancellationToken);
-                }
-
-                this.AddErrorMessage(string.Format(
-                    CultureInfo.CurrentCulture,
-                    Resources.MethodParameterCountDoesNotMatch,
-                    method,
-                    methodParameterCount,
-                    request.ParameterCount));
-
-                return null;
-            }
-
-            // Parameters must be compatible
-            try
-            {
-                return request.GetParameters(method.Parameters, jsonSerializer);
-            }
-            catch (Exception exception)
-            {
-                this.AddErrorMessage(string.Format(CultureInfo.CurrentCulture, Resources.MethodParametersNotCompatible, method, exception.Message));
-                return null;
+                    this.AddErrorMessage(string.Format(
+                        CultureInfo.CurrentCulture,
+                        Resources.MethodParameterCountDoesNotMatch,
+                        method,
+                        methodParameterCount,
+                        request.ArgumentCount));
+                    return false;
+                case JsonRpcRequest.ArgumentMatchResult.ParameterArgumentTypeMismatch:
+                    this.AddErrorMessage(string.Format(CultureInfo.CurrentCulture, Resources.MethodParametersNotCompatible, method));
+                    return false;
+                case JsonRpcRequest.ArgumentMatchResult.MissingArgument:
```

Comment:
FWIW, I would put return false here even though you don't need to, as otherwise it's too easy to add another case below here that does something else